compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt

```diff
@@ -285,28 +289,64 @@ internal class AssistedFactoryGenerator : PrivateCodeGenerator() {
   ) {
     companion object {
       fun FunctionDescriptor.toAssistedFactoryFunction(
+        module: ModuleDescriptor,
         factoryClass: KtClassOrObject
       ): AssistedFactoryFunction {
 
+        val containingClass = containingDeclaration as ClassDescriptor
+
         val returnTypeLazy = lazy(NONE) {
-          returnType
-            ?.classDescriptorForType()
-            ?.fqNameSafe
-            ?: throw AnvilCompilationException(
+
+          fun fail(): Nothing {
+            throw AnvilCompilationException(
               "Invalid return type: ${factoryClass.requireFqName()}. An assisted factory's " +
                 "abstract method must return a type with an @AssistedInject-annotated constructor.",
               element = factoryClass
             )
+          }
+
+          // This can be a generic type like `T` if the function is declared in a super type.
+          // Check that it can resolve before using this type.
+          returnType?.classDescriptorOrNull()
+            ?.fqNameSafe
+            ?: factoryClass.resolveGenericKotlinType(
+              module,
+              containingClass.toClassReference(),
+              returnType ?: fail()
+            )
+              ?.fqNameOrNull(module)
+            ?: fail()
         }
         val parameterPairs = lazy(NONE) {
           valueParameters.map { param ->
-            param.name.asString() to param.type.asTypeName()
+
+            val fromDescriptor = param.type
+              // This can be a generic type like `T` if the function is declared in a super type.
+              // Check that it can resolve before using this type.
+              .takeIf { it.classDescriptorOrNull() != null }
+              ?.asTypeName()
+              ?: factoryClass.resolveGenericKotlinType(
+                module,
+                containingClass.toClassReference(),
+                param.type
+              )
+                ?.requireTypeName(module)
+              ?: run {
+                throw AnvilCompilationException(
+                  "Unable to resolve assisted function's argument type: ${param.type}.",
+                  element = factoryClass
+                )
+              }
```

Comment:
The `run` is redundant here.
```suggestion
              ?: throw AnvilCompilationException(
                "Unable to resolve assisted function's argument type: ${param.type}.",
                element = factoryClass
              )
```

Reply:
Heh...  Thanks!  This was a standalone "fail" function which I inlined via the IDE.  I didn't notice the addition of the `run`.